staging/src/k8s.io/kubectl/pkg/cmd/expose/expose_test.go

```diff
@@ -72,7 +72,7 @@ func TestRunExposeService(t *testing.T) {
 					Selector: map[string]string{"app": "go"},
 				},
 			},
-			expected: "service/foo exposed",
+			expected: "error: selector must be specified",
```

Comment:
I took a lot at how the unit test cases are performed so I think for `TestRunExposeService` it will actually use a `TestFactory` 
```go
tf := cmdtesting.NewTestFactory().WithNamespace(test.ns)
```
to test the code, and it will actually run the `cobra.Command` over here 
```go
cmd.Run(cmd, test.args)
```
So when running the Command, `checkErr` will throw a panic if any error has been found.
```go
cmdutil.CheckErr(o.Complete(f, cmd))
cmdutil.CheckErr(o.Validate())
cmdutil.CheckErr(o.RunExpose(cmd, args))
```

That being said, the unit test `TestRunExposeService` cannot have a test that will return error like the first test case ` "expose-service-from-service-no-selector-defined"`, so you will need to remove that one.

In addition to that, you can also add another test function that perform the validation (you can built on top of the test case) . Take a look at how it can be done similarly over here https://github.com/kubernetes/kubernetes/blob/752c4b7f0b3630fb58f8489cb30748be79cde4f2/staging/src/k8s.io/kubectl/pkg/cmd/create/create_configmap_test.go#L445-L458


Basically you don't need to run the whole command but instead just **create the expose object and compare the error**.
Does that make sense?

Reply:
Hello @lauchokyip ,`no-name-passed-from-the-cli` this error will also be shown after removing first one

How will we send data to validate function do we have to use same test cases?
Do we have to make `exposeServiceOptions` structure and store input of testcases, then check if error we got is same for testcases. also we have to add a new testcase without selector?

Reply:
> no-name-passed-from-the-cli this error will also be shown after removing first one

`no-name-passed-from-the-cli` will fail because when checking the name in `Validate()` it will return an error. 

> How will we send data to validate function do we have to use same test cases?

We don't, you need to remove all the test cases that will fail during `Validate()` and create another test function.

> Do we have to make exposeServiceOptions structure and store input of testcases, then check if error we got is same for testcases. also we have to add a new testcase for checking without selector?

Yes please, that would be great!

Reply:
@almas33 , just found out , it is here (https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubectl/pkg/generate/versioned/service_test.go) You can add the test cases here to the file

Reply:
I have removed some test cases which were there in validate but after that it is giving FAIL although expected value is what it wants but still it is showing FAIL
```
=== RUN   TestRunExposeService
=== RUN   TestRunExposeService/expose-service-from-service
    expose_test.go:432: expose-service-from-service: Unexpected output! Expected
        TestRunExposeService/foo exposed
        got
=== RUN   TestRunExposeService/expose-service
    expose_test.go:432: expose-service: Unexpected output! Expected
        service/foo exposed (dry run)
        got
=== RUN   TestRunExposeService/expose-affinity-service
    expose_test.go:432: expose-affinity-service: Unexpected output! Expected
        service/foo exposed (dry run)
        got
=== RUN   TestRunExposeService/expose-service-cluster-ip
    expose_test.go:432: expose-service-cluster-ip: Unexpected output! Expected
        service/foo exposed (dry run)
        got
=== RUN   TestRunExposeService/expose-headless-service
    expose_test.go:432: expose-headless-service: Unexpected output! Expected
        service/foo exposed (dry run)
        got
=== RUN   TestRunExposeService/expose-headless-service-no-port
    expose_test.go:432: expose-headless-service-no-port: Unexpected output! Expected
        service/foo exposed (dry run)
        got
=== RUN   TestRunExposeService/expose-from-file
    expose_test.go:432: expose-from-file: Unexpected output! Expected
        service/foo exposed (dry run)
        got
=== RUN   TestRunExposeService/expose-service-from-service-no-selector-defined-sctp
error: selector must be specified
exit status 1
FAIL    k8s.io/kubectl/pkg/cmd/expose   0.069s
almas@demon-workshop:~/Documents/kubernetes/staging/src/k8s.io/kubectl/pkg/cmd/expose$ sudo go test -v
=== RUN   TestRunExposeService
=== RUN   TestRunExposeService/expose-service-from-service
    expose_test.go:432: expose-service-from-service: Unexpected output! Expected
        TestRunExposeService/foo exposed
        got
=== RUN   TestRunExposeService/expose-service
    expose_test.go:432: expose-service: Unexpected output! Expected
        service/foo exposed (dry run)
        got
=== RUN   TestRunExposeService/expose-affinity-service
    expose_test.go:432: expose-affinity-service: Unexpected output! Expected
        service/foo exposed (dry run)
        got
=== RUN   TestRunExposeService/expose-service-cluster-ip
    expose_test.go:432: expose-service-cluster-ip: Unexpected output! Expected
        service/foo exposed (dry run)
        got
=== RUN   TestRunExposeService/expose-headless-service
    expose_test.go:432: expose-headless-service: Unexpected output! Expected
        service/foo exposed (dry run)
        got
=== RUN   TestRunExposeService/expose-headless-service-no-port
    expose_test.go:432: expose-headless-service-no-port: Unexpected output! Expected
        service/foo exposed (dry run)
        got
=== RUN   TestRunExposeService/expose-from-file
    expose_test.go:432: expose-from-file: Unexpected output! Expected
        service/foo exposed (dry run)
        got
=== RUN   TestRunExposeService/expose-service-from-service-sctp
    expose_test.go:432: expose-service-from-service-sctp: Unexpected output! Expected
        service/foo exposed
        got
=== RUN   TestRunExposeService/expose-service-cluster-ip-sctp
    expose_test.go:432: expose-service-cluster-ip-sctp: Unexpected output! Expected
        service/foo exposed (dry run)
        got
=== RUN   TestRunExposeService/expose-headless-service-sctp
    expose_test.go:432: expose-headless-service-sctp: Unexpected output! Expected
        service/foo exposed (dry run)
        got
=== RUN   TestRunExposeService/namespace-yaml
    expose_test.go:432: namespace-yaml: Unexpected output! Expected
        namespace: testns
        got
--- FAIL: TestRunExposeService (0.07s)
    --- FAIL: TestRunExposeService/expose-service-from-service (0.01s)
    --- FAIL: TestRunExposeService/expose-service (0.00s)
    --- FAIL: TestRunExposeService/expose-affinity-service (0.01s)
    --- FAIL: TestRunExposeService/expose-service-cluster-ip (0.01s)
    --- FAIL: TestRunExposeService/expose-headless-service (0.01s)
    --- FAIL: TestRunExposeService/expose-headless-service-no-port (0.01s)
    --- FAIL: TestRunExposeService/expose-from-file (0.01s)
    --- FAIL: TestRunExposeService/expose-service-from-service-sctp (0.01s)
    --- FAIL: TestRunExposeService/expose-service-cluster-ip-sctp (0.01s)
    --- FAIL: TestRunExposeService/expose-headless-service-sctp (0.01s)
    --- FAIL: TestRunExposeService/namespace-yaml (0.01s)
FAIL
exit status 1
FAIL    k8s.io/kubectl/pkg/cmd/expose   0.088s
```

Reply:
I have removed some test cases which are to be covered by va;idate() but it is showing FAIL in other cases , I haven't added anything yet ,expected value is same what it wants. can you see what might be the problem
```
=== RUN   TestRunExposeService
=== RUN   TestRunExposeService/expose-service-from-service
    expose_test.go:432: expose-service-from-service: Unexpected output! Expected
        TestRunExposeService/foo exposed
        got
=== RUN   TestRunExposeService/expose-service
    expose_test.go:432: expose-service: Unexpected output! Expected
        service/foo exposed (dry run)
        got
=== RUN   TestRunExposeService/expose-affinity-service
    expose_test.go:432: expose-affinity-service: Unexpected output! Expected
        service/foo exposed (dry run)
        got
=== RUN   TestRunExposeService/expose-service-cluster-ip
    expose_test.go:432: expose-service-cluster-ip: Unexpected output! Expected
        service/foo exposed (dry run)
        got
=== RUN   TestRunExposeService/expose-headless-service
    expose_test.go:432: expose-headless-service: Unexpected output! Expected
        service/foo exposed (dry run)
        got
=== RUN   TestRunExposeService/expose-headless-service-no-port
    expose_test.go:432: expose-headless-service-no-port: Unexpected output! Expected
        service/foo exposed (dry run)
        got
=== RUN   TestRunExposeService/expose-from-file
    expose_test.go:432: expose-from-file: Unexpected output! Expected
        service/foo exposed (dry run)
        got
=== RUN   TestRunExposeService/expose-service-from-service-sctp
    expose_test.go:432: expose-service-from-service-sctp: Unexpected output! Expected
        service/foo exposed
        got
=== RUN   TestRunExposeService/expose-service-cluster-ip-sctp
    expose_test.go:432: expose-service-cluster-ip-sctp: Unexpected output! Expected
        service/foo exposed (dry run)
        got
=== RUN   TestRunExposeService/expose-headless-service-sctp
    expose_test.go:432: expose-headless-service-sctp: Unexpected output! Expected
        service/foo exposed (dry run)
        got
=== RUN   TestRunExposeService/namespace-yaml
    expose_test.go:432: namespace-yaml: Unexpected output! Expected
        namespace: testns
        got
--- FAIL: TestRunExposeService (0.07s)
    --- FAIL: TestRunExposeService/expose-service-from-service (0.01s)
    --- FAIL: TestRunExposeService/expose-service (0.00s)
    --- FAIL: TestRunExposeService/expose-affinity-service (0.01s)
    --- FAIL: TestRunExposeService/expose-service-cluster-ip (0.01s)
    --- FAIL: TestRunExposeService/expose-headless-service (0.01s)
    --- FAIL: TestRunExposeService/expose-headless-service-no-port (0.01s)
    --- FAIL: TestRunExposeService/expose-from-file (0.01s)
    --- FAIL: TestRunExposeService/expose-service-from-service-sctp (0.01s)
    --- FAIL: TestRunExposeService/expose-service-cluster-ip-sctp (0.01s)
    --- FAIL: TestRunExposeService/expose-headless-service-sctp (0.01s)
    --- FAIL: TestRunExposeService/namespace-yaml (0.01s)
FAIL
exit status 1
FAIL    k8s.io/kubectl/pkg/cmd/expose   0.088s
```

Reply:
> @almas33 , just found out , it is here (https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubectl/pkg/generate/versioned/service_test.go) You can add the test cases here to a new test function

generator is not passed here so , I guess it has to be removed

Reply:
Check https://github.com/kubernetes/kubernetes/pull/99891#discussion_r715299971

Looks like you aren't calling the `checkService`  after the if statement `if inline := cmdutil.GetFlagString(cmd, "overrides"); len(inline) > 0`

Reply:
> generator is not passed here so , I guess it has to be removed

Even the generator is not passed over there, you can still **apply the same concept** (pass in the `ExposeServiceOptions` struct  and compare the generated `corev1.ExposeService` `ExposeServiceOptions`  using  with expected `corev1.ExposeService` )

Reply:
> > generator is not passed here so , I guess it has to be removed
> 
> Even the generator is not passed over there, you can still **apply the same concept** (pass in the `ExposeServiceOptions` struct and compare the generated `corev1.ExposeService` `ExposeServiceOptions` using with expected `corev1.ExposeService` )
> 
> Take a look at this please
> 
> https://github.com/kubernetes/kubernetes/blob/752c4b7f0b3630fb58f8489cb30748be79cde4f2/staging/src/k8s.io/kubectl/pkg/cmd/create/create_configmap_test.go#L430-L460
> 
> 
> 
> Similarly you can do something like
> 
> ```go
> test := {
>  ...
> }{
> for name, test := range tests {
> 	t.Run(name, func(t *testing.T) {
> 		var service *corev1.ExposeService = nil
> 		serviceExposeOptions := ServiceExposeOptions{
> 		...
> 		}
>                ...
>                err := serviceExposeOptions.Validate()
> 
> 		if err == nil {
> 			service, err = serviceExposeOptions.createService()
> 		}
> 
>               ...
>               !apiequality.Semantic.DeepEqual(configMap, test.expected) 
>               ...
>               require.EqualError(t, err, test.expectErr)
>    
>       }
> 
> }
> ```

Yes, I saw that , I meant the generator which is written in test cases we can  put just parameters and then we can pass those parameters as a `struct` to get validated. I don't quite understood the use of generator which is present in https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubectl/pkg/generate/versioned/service_test.go

Reply:
not sure if I can explain it well but the presence of `generator` is to help create an object. so before this PR (https://github.com/kubernetes/kubernetes/commit/da19e5b2583fa00a13110de97448bf6452010988) there are actually different generators for different objects (ConfigMap, namespace, secrets, etc etc). 
</br>
To create an object you just need to know what `generator` you want to use, for example `generator["service/v1"]` will return a service V1 generator and the `params` to pass in. So instead of having different function for different object
```go
serviceOptions.createService()
configMapOptions.createConfigMap()
nameSpaceOptions.createNameSpace()
...
```
You just need to know the name
```go
generator["service/v1"].Generate()
generator["configmap/v1"].Generate()
```
The concept is similar to Object Oriented Design, a Dog and a Cat are both Animal so you can do 
```
Animal a = new Dog();
Animal b = new Cat();
a.Sound() // Bark
b.Sound() // Meow
```
Hopefully it makes sense

Reply:
So then first `ExposeServiceOptions` should be implemented by generate then only we can use it

Reply:
> > generator is not passed here so , I guess it has to be removed
> 
> Even the generator is not passed over there, you can still **apply the same concept** (pass in the `ExposeServiceOptions` struct and compare the generated `corev1.ExposeService` `ExposeServiceOptions` using with expected `corev1.ExposeService` )
> 
Is `ExposeService` present in corev1 package?

Reply:
As container-port is not used in expose service, so can I safely remove them while using test cases from this file https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubectl/pkg/generate/versioned/service_test.go

Reply:
Can you add one test case where you if `target-port` is not given it will use `container-port` ? 
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubectl/pkg/generate/versioned/service.go#L191

Reply:
https://github.com/kubernetes/kubernetes/pull/99891/files#r592533964  As this comment says,so for test cases where container port is used should I put there target port instead

Reply:
Good catch, yes like @rikatz said, you should remove it

Reply:
For port default name is `""` in expose.go but for sevice.go it is `"default"` so it should I replace it in expose.go as `"default"` or in expose_test.go as `""`
And protocols are not getting parsed can you see if function is right, for eg. if `8080/UDP,8081/TCP,8082/SCTP` is given as input it will output object as it is for every port and doesn't parse it as UDP , TCP, SCTP